static/app/views/dashboards/dashboardLegendUtils.tsx

```diff
@@ -0,0 +1,204 @@
+import type {Location} from 'history';
+
+import type {Series} from 'sentry/types/echarts';
+import type {InjectedRouter} from 'sentry/types/legacyReactRouter';
+import type {Organization} from 'sentry/types/organization';
+import {decodeList} from 'sentry/utils/queryString';
+
+import {type DashboardDetails, DisplayType, type Widget} from './types';
+
+type Props = {
+  dashboard: DashboardDetails | null;
+  location: Location;
+  organization: Organization;
+  router: InjectedRouter;
+};
+
+class DashboardLegendEncoderDecoder {
+  dashboard: DashboardDetails | null;
+  location: Location;
+  organization: Organization;
+  router: InjectedRouter;
+
+  constructor(props: Props) {
+    this.dashboard = props.dashboard;
+    this.location = props.location;
+    this.organization = props.organization;
+    this.router = props.router;
+  }
+
+  // Updates legend param when a legend selection has been changed
+  updateLegendQueryParam(selected: Record<string, boolean>, widget: Widget) {
```

Comment:
I think `setWidgetSelectionState` is clearer here. Maybe even
```suggestion
  setWidgetSelectionState(widget: Widget, selection: LegendSelection) {
```

And 

```jsx
type LegendSelection = Record<string, boolean>;
```

Somewhere higher up.

This way the class name nicely matches the kind of data that's passed around. `WidgetLegendSelectionState` manages widgets and selections